### DISCUSSION TITLE
Query post view

### DISCUSSION THREAD

**Discussion by IRediTOTO** [wp-graphql/wp-graphql#1860] [Q&A] [answered]
hi, is there anyway help us query post view count? I mean number of view for a post. 
I read this post https://gretathemes.com/count-post-views/
but can you help me custom to export view number to WP-graphql ? Thank you so much

**Comment by jasonbahl** [comment, marked as answer]
@IRediTOTO I know I'm a bit late to responding here, but if you store the "view count" as post meta as described in the article, you should be able to register a field to the GraphQL Schema that can resolve that value. 

If you want the field to exist only on one post type, such as the "post" post type, you could register the field like so: 

```php
add_action( 'graphql_register_types', function() {

  register_graphql_field( 'Post', 'viewCount', [ 
    'type' => 'Int',
    'description' => __( 'The number of times the post was viewed', 'your-textdomain' ),
    'resolve' => function( $post, $args, $context, $info ) {
      // $post is an instance of a WPGraphQL Post model, 
      // so we can get the ID from the post, and then get meta from that post
      // by calling get_post_meta()
      $view_count = get_post_meta( $post->databaseId, 'post_views_count', true );
      
      // if there's any count value, return it, otherwise return 0
      return isset( $view_count ) ? absint( $view_count ) : 0;
    }
  ]);

});
```

We can see this snippet of code in action using GraphiQL here: 

![Screen Shot 2021-12-10 at 2 10 20 AM](https://user-images.githubusercontent.com/1260765/145548111-743b8a29-af46-4646-b91e-182c4f10cf80.png)

**Reply by masitings** [reply]
And how do i update the viewCount ? was that using mutation ? After registering this, do you have an example how to create field on mutations ?

**Reply by LucasTrindadeDev** [reply]
I have the same question. All the tutorials I have seen only shows how to query the post views/count, but not how to update them with wpgraphql. Did you get any updates?

**Reply by metawise** [reply]
Did you guys figure the answer?

**Reply by masitings** [reply]
i think i found the solutions for this .. and i think the solution was the same but more details, put the code below in your **functions.php** theme's file. 
```php

add_action( 'graphql_register_types', function() {
	register_graphql_mutation( 'updatePostView', [
		'inputFields'         => [
			'views' => [
				'type' => 'Int',
				'description' => __( 'Post views', 'wp-graphql' ),
			],
			'slug' => [
				'type' => 'String',
				'description' => __( 'Post Slug', 'wp-graphql' ),
			]
		],

		'outputFields'        => [
			'views' => [
				'type' => 'Number',
				'description' => __( 'Post views', 'wp-graphql' ),
				'resolve' => function( $payload, $args, $context, $info ) {
					return isset( $payload['views'] ) ? $payload['views'] : null;
				}
			]
		],

		'mutateAndGetPayload' => function( $input, $context, $info ) {
            $queried_post = get_page_by_path($input['slug'], OBJECT, 'post');
			update_post_meta( $queried_post->ID, 'post_views_count', $input['views'] );
			$views = get_post_meta( $queried_post->ID, 'post_views_count', true );
			return [
				'views' => $views
			];
		}
	] );

    register_graphql_field( 'Post', 'viewCount', [ 
        'type' => 'Int',
        'description' => __( 'The number of times the post was viewed', 'your-textdomain' ),
        'resolve' => function( $post, $args, $context, $info ) {
            $view_count = get_post_meta( $post->databaseId, 'post_views_count', true );
            return isset( $view_count ) ? absint( $view_count ) : 0;
        }
    ]);
});
```
The idea behind this is to add a new object called viewCount and make it be able to be updated using mutation. When you visit your Graphql Playground, you'll see a mutation called `updatePostView` where you called everytime user visit your post.

**Reply by AliNaraghiA** [reply]
@masitings are you sure this code works well?
 I can't find Sth for showing popular-posts with Wpgraphql.

**Reply by dotwise** [reply]
Note that this mutation doesn't actually increase the post view count. It just sets it to whatever is the value of $input["views"], which is how it is meant to work I think.

If you want to have the mutation increment the post view count itself you could change it to something like this (please note my code uses the post id instead of the post slug as @masitings code does, to retrieve the post in question.

```
    'mutateAndGetPayload' => function ($input, $context, $info) {

        # perform the necessary database work
        $queried_post = get_post($input['postId']);

        $views = get_post_meta($queried_post->ID, 'views', true);
        $views++;
        update_post_meta($queried_post->ID, 'views', $views);

        return [
            'views' => $views,
        ];
    },
```

**Reply by masitings** [reply]
Sure do, i've using this at my site. But in order to do the increment of view, you need to call the mutation and increase it manually by getting the count of view of the post.  @AliNaraghiA
